Invalid C code can be generated when using std/sugar arrow syntax without specifying types

### What happened?

Minimal reproducible example.
```
import std/sequtils
import std/sugar

proc test() =
  var xs, ys: seq[int]
  for i in 0..5:
    xs.add(i)

  xs.apply(d => ys.add(d))
  assert ys == @[0, 1, 2, 3, 4, 5]

test()
```

This generates the gcc (v 12.2.0) type error below.

### Nim Version

Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-10-30
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: f8b5464f31834bf97e7f9023644e255f2e9b6485
active boot switches: -d:release

### Current Standard Output Logs

```shell
CC: example.nim
/home/mike/.cache/nim/example_d/@mexample.nim.c: In function ‘test__example_3’:
/home/mike/.cache/nim/example_d/@mexample.nim.c:344:48: error: incompatible type for argument 3 of ‘apply__example_106’
  344 |         apply__example_106(xs.p->data, xs.len, T5_);
      |                                                ^~~
      |                                                |
      |                                                tyProc__SqRJqa5aOiKB4EuKlKh9asA
/home/mike/.cache/nim/example_d/@mexample.nim.c:179:93: note: expected ‘tyProc__vwnqUR9cTZ8xwFKGcxd2hkw’ but argument is of type ‘tyProc__SqRJqa5aOiKB4EuKlKh9asA’
  179 | static N_INLINE(void, apply__example_106)(NI* s, NI sLen_0, tyProc__vwnqUR9cTZ8xwFKGcxd2hkw op) {
```


### Expected Standard Output Logs

```shell
I would expect this to either work or give a type mismatch error indicating the line with the problem.
```


### Possible Solution

_No response_

### Additional Information

If this same code is at top level it compiles and runs fine.

If `xs` is initialized on a single line such as `var xs = @[0, 1, 2, 3, 4, 5]` then you get a proper type mismatch error from nim during compilation. Oddly(?), when initialized that way it also causes a type mismatch error when run at top level.

Not 100% sure, but I believe that without arrow syntax (anonymous proc without std/sugar or do syntax) that the type is always required so this can't happen.

--- a/tests/stdlib/tsugar.nim
+++ b/tests/stdlib/tsugar.nim
@@ -3,7 +3,7 @@ discard """
 x + y = 30
 '''
 """
-import std/[sugar, algorithm, random, sets, tables, strutils]
+import std/[sugar, algorithm, random, sets, tables, strutils, sequtils]
 import std/[syncio, assertions]
 
 type # for capture test, ref #20679
@@ -286,6 +286,17 @@ proc mainProc() =
     let s2 = dumpToString:
       doAssertRaises(AssertionDefect): doAssert false
     doAssert "except AssertionDefect" in s2
+
+  block: # bug #20704
+    proc test() =
+      var xs, ys: seq[int]
+      for i in 0..5:
+        xs.add(i)
+
+      xs.apply(d => ys.add(d))
+      doAssert ys == @[0, 1, 2, 3, 4, 5]
+
+    test()
 
 static:
   main()

